atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/internalschema/TransactionSchemaManager.java

```diff
@@ -45,33 +48,49 @@ public int getTransactionsSchemaVersion(long timestamp) {
-                    .stream()
-                    .filter(valueAndBound -> valueAndBound.bound() >= timestamp)
-                    .findAny(),
+                            .stream()
+                            .filter(valueAndBound -> valueAndBound.bound() >= timestamp)
+                            .findAny(),
                     timestamp);
         }
         return possibleVersion.get();
     }
 
     /**
-     * Installs a new transactions table schema version, by submitting a relevant transform.
+     * Attempts to install a new transactions table schema version, by submitting a relevant transform.
+     *
+     * The execution of this method does not guarantee that the provided version will eventually be installed.
+     * This method returns true if and only if in the map agreed by the coordination service evaluated at the validity
+     * bound, the transactions schema version is equal to newVersion.
      */
-    public void tryInstallNewTransactionsSchemaVersion(int newVersion) {
-        coordinationService.tryTransformCurrentValue(valueAndBound -> {
-            if (!valueAndBound.value().isPresent()) {
-                throw new SafeIllegalStateException("Persisted value is empty, which is unexpected.");
-            }
+    public boolean tryInstallNewTransactionsSchemaVersion(int newVersion) {
+        List<Integer> presentVersionPeakValidity = coordinationService.tryTransformCurrentValue(
+                valueAndBound -> installNewVersionInMapOrDefault(newVersion, valueAndBound))
+                .existingValues()
+                .stream()
+                .map(valueAndBound -> valueAndBound.value()
+                        .orElseThrow(() -> new SafeIllegalStateException("Unexpectedly found no value in store"))
+                        .timestampToTransactionsTableSchemaVersion()
+                        .getValueForTimestamp(valueAndBound.bound()))
+                .collect(Collectors.toList());
+        return Iterables.getOnlyElement(presentVersionPeakValidity) == newVersion;
+    }
+
+    private InternalSchemaMetadata installNewVersionInMapOrDefault(int newVersion,
+            ValueAndBound<InternalSchemaMetadata> valueAndBound) {
+        if (!valueAndBound.value().isPresent()) {
+            return InternalSchemaMetadata.defaultValue();
+        }
 
-            InternalSchemaMetadata internalSchemaMetadata = valueAndBound.value().get();
-            return InternalSchemaMetadata.builder()
-                    .from(internalSchemaMetadata)
-                    .timestampToTransactionsTableSchemaVersion(
-                            installNewVersionInMap(
-                                    internalSchemaMetadata.timestampToTransactionsTableSchemaVersion(),
-                                    valueAndBound.bound() + 1,
-                                    newVersion))
-                    .build();
-        });
+        InternalSchemaMetadata internalSchemaMetadata = valueAndBound.value().get();
```

Comment:
InternalSchemaMetadata internalSchemaMetadata = valueAndBound.value().orElse(InternalSchemaMetadata.defaultValue())

Reply:
Yep, that's better. This does change behaviour (in that on the first update you'd previously just install default, while now you install default _and_ install a new version at the right place), but I think it's a good change.

Reply:
Subsequent offline discussion - unfortunately doesn't work because the transform on its first go has no idea what to set the new version of the bound to be. This can be addressed by calling the timestamp service directly, or making transforms a bi-function that also has a timestamp that is the bound we'll propose our value at. For now we'll just use the old semantics...